kvstore/badger/badger.go

```diff
@@ -148,29 +153,84 @@ func (s *badgerStore) DeletePrefix(prefix kvstore.KeyPrefix) error {
 
 func (s *badgerStore) Batched() kvstore.BatchedMutations {
 	return &batchedMutations{
-		batched:  s.instance.NewWriteBatch(),
-		dbPrefix: s.dbPrefix,
+		store:            s.instance,
+		dbPrefix:         s.dbPrefix,
+		setOperations:    make(map[string]kvstore.Value),
+		deleteOperations: make(map[string]types.Empty),
 	}
 }
 
 // batchedMutations is a wrapper around a WriteBatch of a BadgerDB.
 type batchedMutations struct {
-	batched  *badger.WriteBatch
-	dbPrefix []byte
+	store            *badger.DB
+	dbPrefix         []byte
+	setOperations    map[string]kvstore.Value
+	deleteOperations map[string]types.Empty
+	operationsMutex  sync.Mutex
 }
 
 func (b *batchedMutations) Set(key kvstore.Key, value kvstore.Value) error {
-	return b.batched.Set(append(b.dbPrefix, key...), value)
+	stringKey := b.stringKey(key)
+
+	b.operationsMutex.Lock()
+	defer b.operationsMutex.Unlock()
+
+	delete(b.deleteOperations, stringKey)
+	b.setOperations[stringKey] = value
+
+	return nil
 }
 
 func (b *batchedMutations) Delete(key kvstore.Key) error {
-	return b.batched.Delete(append(b.dbPrefix, key...))
+	stringKey := b.stringKey(key)
+
+	b.operationsMutex.Lock()
+	defer b.operationsMutex.Unlock()
+
+	delete(b.setOperations, stringKey)
+	b.deleteOperations[stringKey] = types.Void
+
+	return nil
 }
 
 func (b *batchedMutations) Cancel() {
-	b.batched.Cancel()
+	b.operationsMutex.Lock()
+	defer b.operationsMutex.Unlock()
+
+	b.setOperations = make(map[string]kvstore.Value)
+	b.deleteOperations = make(map[string]types.Empty)
 }
 
 func (b *batchedMutations) Commit() error {
-	return b.batched.Flush()
+	writeBatch := b.store.NewWriteBatch()
+
+	b.operationsMutex.Lock()
+	defer b.operationsMutex.Unlock()
+
+	for key, value := range b.setOperations {
+		err := writeBatch.Set(typeutils.StringToBytes(key), value)
```

Comment:
That is not correct.

Key is a string an not a pointer type. This means that the memory is reallocated for every iteration and we essentially just transform the variable pointer of the current string to a byte slice pointer so it doesn't get garbage collected. This is 100% safe.

See the following example: https://play.golang.org/p/BxVufhxsY8m (Note how value is the same in the goroutines but the transformed byte slices are different for every iteration). Since this is database related stuff we should really make sure it runs fast and use this little trick.

Reply:
You are wrong! The compiler does *a lot* of optimizations under the hood! Especially for maps and strings. And since strings in go are immutable there is a lot of allocations the compiler can avoid.

Check for example the following:
```go
func main() {
	m := map[string]int{
		"1000000000000000000000000000000000000000000000000": 1,
		"2000000000000000000000000000000000000000000000000": 2,
	}

	var x int
	n := testing.AllocsPerRun(100, func() {
		for key, _ := range m {
			x += len(key)
		}
	})
	fmt.Println(n)

	buf := []byte("1000000000000000000000000000000000000000000000000")
	n = testing.AllocsPerRun(100, func() {
		x += m[string(buf)]
	})
	fmt.Println(n)
}
```

You will see that no allocation whatsoever (and therefore no GC) happens in any `func` because the compiler uses the backed byte array for both. (Also see https://github.com/golang/go/commit/f5f5a8b6209f84961687d993b93ea0d397f5d5bf for example.)
Thus `err := writeBatch.Set(typeutils.StringToBytes(key), value)` uses the same underlying memory which is then passed to badger, where it is _not_ copied. During certain map operations this will also replace the memory in badger causing very nasty race-conditions (even though `batchedMutations` is not supposed to be modified after the `Commit` this is still never prevented in code).

Even if in that particular situation, with the current badger version and with the current implementation of maps, such a scenario does not seem to happen, it is still reckless to use `StringToBytes` as the situation or the compiler might change.
This is a very dangerous an unnecessary micro optimization that will just potentially lead to hours of debugging in the future!